src/Tests/Tests/Test_FastConv.cpp

```diff
@@ -18,22 +19,262 @@ namespace fastconv_test {
         }
 
         virtual void TearDown()
         {
+            delete m_pCFastConv;
+            delete[] IR;
         }
 
-        CFastConv *m_pCFastConv = 0;
+        int IRLength = 0;
+        float* IR = nullptr;
+        CFastConv *m_pCFastConv = nullptr;
     };
 
-    TEST_F(FastConv, EmptyTest)
+    TEST_F(TimeDomainConv, identity)
+    {
+        // initialize input signal, which is $\delta[n-3]$
+        int shift = 3;
+        int signalLength = 10;
+        float* input = new float[signalLength]; // make the input signal 10 samples long
+        memset(input, 0, sizeof(float) * signalLength); // an impulse as input signal at sample index 3
+        input[shift] = 1.F;
+
+        // initialize output signal
+        float* output = new float[signalLength];
+        memset(output, 0, sizeof(float) * signalLength);
+
+        m_pCFastConv->process(output, input, signalLength);
+
+        // check if the output signal is exactly the same with the impulse response
+        for (int i = 0; i + shift < signalLength; ++i) {
+            EXPECT_NEAR(IR[i], output[i + shift], 1e-8);
+        }
+
+        // also check if the first three are ok
+        for (int i = 0; i < shift; ++i) {
+            EXPECT_EQ(output[i], 0);
+        }
+
+        delete[] input;
+        delete[] output;
+    }
+
+    TEST_F(TimeDomainConv, flushbuffer) {
+        // initialize input signal, which is $\delta[n-3]$
+        int shift = 3;
+        int signalLength = 10;
+        float* input = new float[signalLength]; // make the input signal 10 samples long
+        memset(input, 0, sizeof(float) * signalLength); // an impulse as input signal at sample index 3
+        input[shift] = 1.F;
+
+        // initialize output signal
+        float* output = new float[IRLength];
+        memset(output, 0, sizeof(float) * IRLength);
+
+        // process & discard result
+        m_pCFastConv->process(output, input, signalLength);
+        memset(output, 0, sizeof(float) * IRLength);
+
+        // flush
+        m_pCFastConv->flushBuffer(output);
+
+        // check if the tail is correct
+        for (int i = 0; i + signalLength - shift < IRLength; ++i) {
+            EXPECT_EQ(output[i], IR[i + signalLength - shift]);
+        }
+
+        delete[] input;
+        delete[] output;
+    }
+
+    TEST_F(TimeDomainConv, varyingBlockSize) {
+        // initialize input signal, which is $\delta[n-3]$
+        int shift = 3;
+        int signalLength = 10000;
+        float* input = new float[signalLength]; // make the input signal 10 samples long
+        memset(input, 0, sizeof(float) * signalLength); // an impulse as input signal at sample index 3
+        input[shift] = 1.F;
+
+        // initialize output signal
+        float* output = new float[signalLength];
+        memset(output, 0, sizeof(float) * signalLength);
+
+        // convolve with varying block size
+        int blockLengths[] = {
+            1, 13, 1023, 2048, 1, 17, 5000, 1897
```

Comment:
With an impulse response of length 51 and an input signal that's all zeros after sample 3, nearly all of the output will simply be zero - so most of the block lengths are not fully tested. (I fear this is masking a bug in the freq domain version.)

I'd suggest swapping the generation of input and IR for this test, since the input is much longer: randomly generate the input signal, and have a simple shifted impulse as the IR. Then the output will be non-zero across all the blocks and can be more meaningfully tested.

Reply:
That makes sense. I have intensified the test and it also passed, see c032a52adc9eab33633cc47cb46cda3064ccfa8d.

I didn't read your second paragraph until finishing this commit (I'm sorry...) and swapping `input` and `IR` seems to be better than current implementation! You are so genius, really.